File: practice/mathematics/fundamentals/k_candy_store/k_candy_store.hs
module Main where

import Control.Monad (forM_)
import qualified Data.Vector as V

triangle :: V.Vector (V.Vector Int) -> Int -> V.Vector (V.Vector Int)
triangle tr 0 = tr
triangle tr n = triangle (V.snoc tr r) (n - 1)
    where
        h  = V.last tr
        r' = V.fromList [((h V.! k) + (h V.! (k + 1))) `mod` 1000000000 | k <- [0..(V.length h - 2)]]
        r  = V.snoc (V.cons 1 r') 1

main :: IO ()
main = do
    t <- readLn :: IO Int
    let pascal = triangle (V.fromList [V.fromList [1, 1]]) 2000
    
    forM_ [1..t] $ \_ -> do
        n <- readLn :: IO Int
        k <- readLn :: IO Int
        print $ (pascal V.! (k + n - 2)) V.! (n - 1)
    

File: practice/mathematics/fundamentals/k_candy_store/k_candy_store.cpp
#include <vector>
#include <iostream>

using namespace std;

const long long MOD = 1000000000;

vector<vector<long long>> triangle(int i_n)
{
    vector<vector<long long>> pascal;
    pascal.push_back(vector<long long>(2, 1));
    
    for (int cnt = 0; cnt < i_n; ++cnt)
    {
        vector<long long> b(1, 1);
        vector<long long> last = pascal.back();
        for (int pos = 0; pos < last.size() - 1; ++pos)
        {
            long long d = last[pos] + last[pos + 1];
            d %= MOD;
            b.push_back(d);
        }
        b.push_back(1);
        
        pascal.push_back(b);
    }
    
    return pascal;
}

int main()
{
    int t, n, k;
    vector<vector<long long>> tr = triangle(2000);
    
    cin >> t;
    while (t--)
    {
        cin >> n >> k;
        cout << tr[k + n - 2][n - 1] << endl;
    }
    return 0;
}

